### PR TITLE
Add Metrics API and RESTAPI class

### PR DESCRIPTION
**Pull request by cgcardona**
* Add `MetricsAPI` class
* Add `RESTAPI` class
* Add tests

Callable like this:

```js
const avalanche: Avalanche = new Avalanche(ip, port, protocol, networkID);
const metrics: Metrics = avalanche.Metrics();
const main = async () => {
  let m = await metrics.getMetrics();
}
```

This addresses [Issue 39](https://github.com/ava-labs/avalanche.js/issues/39).

### CODE REVIEW

File: src/utils/types.ts [hunk 1]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
     }
 }
 
+export class RAWAPI extends APIBase {
```
Comment (collincusce) [inline src/utils/types.ts:80]: This has a lot of JSON RPC stuff in it still but we don't really need that for RAWAPI.

File: src/utils/types.ts [hunk 2]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
...
+                }
+                return resp;
+            }
+        });
+    }   
+
+    /**
+     * Returns the rpcid, a strictly-increasing number, starting from 1, indicating the next request ID that will be sent.
+     */
+    getRPCID = ():number => {
```
Comment (collincusce) [inline src/utils/types.ts:121]: Definitely don't need this function.

File: src/utils/types.ts [hunk 3]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
...
+        rpc["method"] = method;
+
+        // Set parameters if exists
+        if(params){
+            rpc['params'] = params;
+        } else if(this.jrpcVersion == "1.0"){
+            rpc["params"] = [];
+        }
+
+        if(this.jrpcVersion != "1.0") {
```
Comment (collincusce) [inline src/utils/types.ts:96]: No RPC versions required.

File: src/utils/types.ts [hunk 4]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
     }
 }
 
+export class RAWAPI extends APIBase {
+    protected jrpcVersion:string = "2.0";
```
Comment (collincusce) [inline src/utils/types.ts:81]: same here this can go

File: src/utils/types.ts [hunk 5]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
...
+    protected rpcid = 1;
```
Comment (collincusce) [inline src/utils/types.ts:82]: and this

File: src/utils/types.ts [hunk 6]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
...
+
+    /**
+     * 
+     * @param core Reference to the Avalanche instance using this endpoint
+     * @param baseurl Path of the APIs baseurl - ex: "/ext/bc/avm"
+     * @param jrpcVersion The jrpc version to use, default "2.0".
+     */
+    constructor(core:AvalancheCore, baseurl:string, jrpcVersion:string = "2.0") {
+        super(core, baseurl);
+        this.jrpcVersion = jrpcVersion;
```
Comment (collincusce) [inline src/utils/types.ts:133]: etc etc etc

File: src/utils/types.ts [hunk 7]
```diff
@@ -77,6 +77,64 @@ export abstract class APIBase {
     }
 }
 
+export class RAWAPI extends APIBase {
+    protected jrpcVersion:string = "2.0";
+    protected rpcid = 1;
+    callMethod = async (method:string, params?:Array<object> | object, baseurl?:string):Promise<RequestResponseData> => {
```
Comment (collincusce) [inline src/utils/types.ts:83]: "callMethod" ceases to make sense for RAWAPI. Should just be "post", "put", "patch", "delete", or "update"...

actually wtf is RAWAPI actually RESTAPI? Let's turn RAWAPI into RESTAPI instead!!!

File: src/utils/types.ts [hunk 8]
```diff
@@ -77,6 +77,226 @@ export abstract class APIBase {
...
+    }
+
+    /**
+     * 
+     * @param core Reference to the Avalanche instance using this endpoint
+     * @param baseurl Path of the APIs baseurl - ex: "/ext/bc/avm"
+     * @param contentType Determines the type of the entity attached to the incoming request
+     * @param acceptType Determines what type of representation is desired at the client side
+     */
+    constructor(core:AvalancheCore, baseurl:string, contentType:string = "application/json;charset=UTF-8", acceptType:string = "application/json") {
```
Comment (collincusce) [inline src/utils/types.ts:293]: Do not default acceptType as it's optional

File: src/utils/types.ts [hunk 9]
```diff
@@ -77,6 +77,226 @@ export abstract class APIBase {
...
+        };
+        let axConf:AxiosRequestConfig = {
+            baseURL:this.core.getProtocol()+"://"+this.core.getIP()+":"+this.core.getPort(),
+            responseType: 'json'
+        };
+
+        return this.core.get(ep, {}, headers, axConf).then((resp:RequestResponseData) => {
+            if(resp.status >= 200 && resp.status < 300){
+                if(typeof resp.data === 'object' && 'error' in resp.data) {
+                    throw new Error("Error returned: " + JSON.stringify(resp.data));
```
Comment (collincusce) [inline src/utils/types.ts:109]: let's remove the throw logic and just return the response. HTTP status codes like 404's have logical meaning in REST. See this guide: https://restfulapi.net/http-status-codes/

File: src/utils/types.ts [hunk 10]
```diff
@@ -77,6 +77,226 @@ export abstract class APIBase {
...
+export class RESTAPI extends APIBase {
+    protected contentType:string;
+    protected acceptType:string;
+
+    get = async (baseurl?:string, contentType?:string, acceptType?:string):Promise<RequestResponseData> => {
+        let ep:string = baseurl ? baseurl : this.baseurl;
+
+        // Set contentType if exists
+        if(contentType) {
+            this.contentType = contentType;
```
Comment (collincusce) [inline src/utils/types.ts:89]: the get function shouldn't overwrite the default value defined in the function. This error is in post, put, delete, patch as well.

File: src/utils/types.ts [hunk 11]
```diff
@@ -77,6 +77,226 @@ export abstract class APIBase {
...
+        }
+
+        // Set acceptType if exists
+        if(acceptType) {
+            this.acceptType = acceptType;
```
Comment (collincusce) [inline src/utils/types.ts:94]: the get function shouldn't overwrite the default value defined in the function. This error is in post, put, delete, patch as well.

File: src/utils/types.ts [hunk 12]
```diff
@@ -77,6 +77,226 @@ export abstract class APIBase {
...
+        }
+
+        let headers:object = {
+            "Content-Type": this.contentType,
+            "Accept": this.acceptType
```
Comment (collincusce) [inline src/utils/types.ts:99]: Accept is optional.

File: tests/apis/avm/utxos.test.ts [hunk 13]
```diff
@@ -79,14 +79,12 @@ let setMergeTester = (input:UTXOSet, equal:Array<UTXOSet>, notEqual:Array<UTXOSe
     let instr:string = JSON.stringify(input.getUTXOIDs().sort());
     for(let i:number = 0; i < equal.length; i++){
         if(JSON.stringify(equal[i].getUTXOIDs().sort()) != instr){
-            console.log("equal failed: ", input.getUTXOIDs().sort(), i, equal[i].getUTXOIDs().sort());
```
Comment (collincusce) [inline tests/apis/avm/utxos.test.ts:82]: Just curious why this log was removed. I guess it's fine to remove it.

Comment (cgcardona) [reply tests/apis/avm/utxos.test.ts:82]: Just cleaning up.

File: src/utils/types.ts [hunk 14]
```diff
@@ -81,39 +81,25 @@ export class RESTAPI extends APIBase {
     protected contentType:string;
     protected acceptType:string;
 
-    get = async (baseurl?:string, contentType?:string, acceptType?:string):Promise<RequestResponseData> => {
+    get = async (baseurl?:string):Promise<RequestResponseData> => {
         let ep:string = baseurl ? baseurl : this.baseurl;
 
-        // Set contentType if exists
```
Comment (collincusce) [inline src/utils/types.ts]: Please don't remove the params, that's not what I asked for. I said don't overwrite the default value. That doesn't mean the params were wrong.

File: src/utils/types.ts [hunk 15]
```diff
@@ -81,39 +81,25 @@ export class RESTAPI extends APIBase {
...
-        if(contentType) {
-            this.contentType = contentType;
+        let headers:object = { "Content-Type": this.contentType };
+        if(this.acceptType) {
```
Comment (collincusce) [inline src/utils/types.ts:88]: You'll need to set an intermediate variable. 
let acceptStr = this.acceptType;
If (acceptType){
acceptStr = acceptType
}

**Pull request merged**